src/test/java/com/artipie/asto/StorageExtension.java

```diff
@@ -0,0 +1,174 @@
+import org.junit.jupiter.api.extension.ParameterContext;
+import org.junit.jupiter.api.extension.ParameterResolutionException;
+import org.junit.jupiter.api.extension.ParameterResolver;
+import org.junit.jupiter.api.extension.TestTemplateInvocationContext;
+import org.junit.jupiter.api.extension.TestTemplateInvocationContextProvider;
+import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
+import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.s3.S3AsyncClient;
+import software.amazon.awssdk.services.s3.model.CreateBucketRequest;
+
+/**
+ * JUnit5 test template extension running test methods with all Storage types.
+ *
+ * @since 0.15
+ */
+final class StorageExtension
+    implements TestTemplateInvocationContextProvider, BeforeAllCallback, AfterAllCallback {
+
+    /**
+     * S3 mock server extension.
+     */
+    private final S3MockExtension mock = S3MockExtension.builder()
+        .withSecureConnection(false)
+        .build();
+
+    @Override
+    public void beforeAll(final ExtensionContext extension) {
+        this.mock.beforeAll(extension);
+    }
+
+    @Override
+    public void afterAll(final ExtensionContext extension) {
+        this.mock.afterAll(extension);
+    }
+
+    @Override
+    public boolean supportsTestTemplate(final ExtensionContext context) {
+        return context.getTestMethod().isPresent();
+    }
+
+    @Override
+    public Stream<TestTemplateInvocationContext> provideTestTemplateInvocationContexts(
+        final ExtensionContext context) {
+        final Collection<Storage> storages;
+        try {
+            storages = Arrays.asList(
+                new InMemoryStorage(),
+                new InMemoryStorage().transaction(Collections.emptyList()).get(),
+                this.s3Storage()
+            );
+        } catch (final InterruptedException | ExecutionException ex) {
+            throw new IllegalStateException("Failed to generate storage", ex);
+        }
+        return storages.stream().map(StorageContext::new);
+    }
+
+    /**
+     * Creates {@link S3Storage} instance with mock endpoint and random bucket.
+     *
+     * @return S3Storage instance.
+     * @checkstyle MethodNameCheck (2 line)
+     */
+    private S3Storage s3Storage() {
+        final S3AsyncClient client = S3AsyncClient.builder()
+            .region(Region.of("us-east-1"))
+            .credentialsProvider(
+                StaticCredentialsProvider.create(AwsBasicCredentials.create("foo", "bar"))
+            )
+            .endpointOverride(
+                URI.create(String.format("http://localhost:%d", this.mock.getHttpPort()))
+            )
+            .build();
+        final String bucket = UUID.randomUUID().toString();
+        final S3Storage storage = new S3Storage(client, bucket);
+        client.createBucket(CreateBucketRequest.builder().bucket(bucket).build()).join();
```

Comment:
@olegmoz Could we create the bucket before instantiating the storage, and then just return the `S3Storage` constructor result? (so without creating `storage` variable?)

Reply:
@HDouss thank you, this way it is cleaner, fixed